File: typescript/test/screenIndex.test.ts
import { readdirSync, readFileSync } from "node:fs";
import { describe, expect, it } from "vitest";
import {
  buildIndex,
  type IndexInput,
  renderIndex,
  searchIndex,
  sizeOf,
} from "../src/index.js";

const dir = "../spec/examples";
const shipped: IndexInput[] = readdirSync(dir)
  .filter((file) => file.endsWith(".yaml"))
  .map((file) => ({ file, source: readFileSync(`${dir}/${file}`, "utf8") }));

const index = buildIndex(shipped);

const PAGE = `page:
  type: crud
  id: order_list
  title: 受注一覧
  repository: orderRepository
  key: orderNo
  search:
    filters:
      - { field: customer, label: 得意先 }
  table:
    columns:
      - { field: orderNo, label: 受注番号 }
      - { field: amount, label: 金額, type: number, format: currency }
  form:
    sections:
      - fields:
          - { field: orderNo, label: 受注番号, required: true }
  actions:
    - { id: csv, type: export, label: CSV出力 }
`;

describe("画面の索引", () => {
  it("app の中身も1枚ずつ数える（ファイル単位ではなく画面単位）", () => {
    // 同梱は11ファイル。うち sales_app が8枚を持つので、画面は18枚になる。
    expect(shipped).toHaveLength(11);
    expect(index.screens).toHaveLength(18);
    expect(index.unreadable).toEqual([]);
  });

  it("1行の要約をそのまま持つ（索引のために別の語彙を作らない）", () => {
    const one = buildIndex([{ file: "page.yaml", source: PAGE }]).screens[0];
    expect(one.brief).toBe(
      "受注一覧（order_list）… 検索＋一覧＋登録・修正・削除。条件 1、列 2、項目 1（必須 1）、ボタン 1、orderRepository から",
    );
    expect(one.what).toBe("検索＋一覧＋登録・修正・削除");
    expect(one.counts).toEqual({
      filters: 1,
      columns: 2,
      sections: 1,
      fields: 1,
      required: 1,
      actions: 1,
    });
  });

  it("並びは「ファイル → id」で固定（同じ入力なら同じ索引）", () => {
    const again = buildIndex(shipped);
    expect(again.screens.map((s) => `${s.file}/${s.id}`)).toEqual(
      index.screens.map((s) => `${s.file}/${s.id}`),
    );
  });

  it("現場の言葉でも実装の言葉でも当たる", () => {
    const one = buildIndex([{ file: "page.yaml", source: PAGE }]);
    // ラベル（現場）と項目名（実装）の両方が語に入っている。
    expect(searchIndex(one, "得意先")).toHaveLength(1);
    expect(searchIndex(one, "customer")).toHaveLength(1);
    expect(searchIndex(one, "orderRepository")).toHaveLength(1);
    expect(searchIndex(one, "CSV出力")).toHaveLength(1);
  });

  it("語の AND で絞る", () => {
    expect(searchIndex(index, "顧客 検索").length).toBeGreaterThan(0);
    // 両方を含む画面だけ。片方しか無いものは落ちる。
    const both = searchIndex(index, "顧客 帳票");
    expect(both).toEqual([]);
  });

  it("種別は「検索」のような現場の言葉でも当たる（説明の語彙を使う）", () => {
    // master の見出し語は「マスタ保守」だが、説明の語彙は「検索・一覧・登録…」なので
    // 「検索」で探しても出る。
    const masters = searchIndex(index, "マスタ 検索").map((s) => s.kind);
    expect(masters).toContain("master");
  });

  it("当たらなければ空", () => {
    expect(searchIndex(index, "ブロックチェーン")).toEqual([]);
    expect(renderIndex([])).toContain("当てはまる画面はありません");
  });

  it("語が無ければ全件（並びはそのまま）", () => {
    expect(searchIndex(index)).toHaveLength(index.screens.length);
    expect(searchIndex(index, "   ")).toHaveLength(index.screens.length);
  });

  it("規模で並べ替えられる（大きい画面から見たいとき）", () => {
    const sorted = [...index.screens].sort((a, b) => sizeOf(b) - sizeOf(a));
    expect(sizeOf(sorted[0])).toBeGreaterThan(sizeOf(sorted[sorted.length - 1]));
    // 必須の数は規模に数えない（項目数に含まれているので二重に効く）。
    expect(
      sizeOf({
        ...index.screens[0],
        counts: { fields: 2, required: 2, controlled: 2 },
      }),
    ).toBe(2);
  });

  it("定義でないファイルは飛ばし、読めない定義は言う（索引は消さない）", () => {
    const result = buildIndex([
      { file: "pubspec.yaml", source: "name: demo\n" },
      { file: "broken.yaml", source: "page:\n  type: crud\n" }, // id / title が無い
      { file: "page.yaml", source: PAGE },
    ]);
    expect(result.ignored).toBe(1);
    expect(result.unreadable.map((one) => one.file)).toEqual(["broken.yaml"]);
    expect(result.screens).toHaveLength(1);
  });

  it("綴り間違いのある定義も索引に載せる（消すと余計に探せない）", () => {
    const typo = PAGE.replace("label: 金額", "label: 金額, witdh: 140");
    const result = buildIndex([{ file: "page.yaml", source: typo }]);
    expect(result.screens).toHaveLength(1);
    expect(result.unreadable).toEqual([]);
  });

  it("人が読む形は画面名の桁が揃う（目で追える表になる）", () => {
    const screens = index.screens.slice(0, 3);
    const lines = renderIndex(screens, { showFile: false }).split("\n").slice(1);
    // 画面名がどの行でも同じ桁から始まる（id の長さがバラバラでも）。
    const columns = lines.map((line, at) => line.indexOf(screens[at].title));
    expect(new Set(columns).size).toBe(1);
    expect(columns[0]).toBeGreaterThan(0);
  });
});

File: flutter/packages/hatake_core/test/screen_index_test.dart
import 'package:hatake_core/hatake_core.dart';
import 'package:test/test.dart';

/// The index over pages that are already in memory — the case that matters
/// inside an app (a screen picker, a "jump to screen" box). Reading definitions
/// off disk is hatake_yaml's job and is tested there.
void main() {
  const orderSearch = SearchPageDefinition(
    id: 'order_search',
    title: '受注照会',
    repository: 'orderRepository',
    keyField: 'orderNo',
    search: SearchDefinition(
      filters: [
        FilterDefinition(field: 'orderNo', label: '受注番号'),
        FilterDefinition(
          field: 'customer',
          label: '顧客名',
          operator: FilterOperators.contains,
        ),
      ],
    ),
    table: TableDefinition(
      columns: [
        ColumnDefinition(field: 'orderNo', label: '受注番号', sortable: true),
        ColumnDefinition(field: 'amount', label: '金額', format: 'currency'),
      ],
    ),
    actions: [
      ActionDefinition(id: 'csv', type: ActionTypes.export, label: 'CSV出力'),
    ],
  );

  const customerMaster = MasterPageDefinition(
    id: 'customer_master',
    title: '顧客マスタ',
    repository: 'customerRepository',
    table: TableDefinition(
      columns: [ColumnDefinition(field: 'code', label: 'コード')],
    ),
    form: FormDefinition(
      sections: [
        SectionDefinition(
          fields: [
            FieldDefinition(field: 'code', label: 'コード', required: true),
            FieldDefinition(field: 'name', label: '顧客名'),
          ],
        ),
      ],
    ),
  );

  const wizard = WizardPageDefinition(
    id: 'customer_wizard',
    title: '顧客登録ウィザード',
    repository: 'customerRepository',
    steps: [
      WizardStepDefinition(
        id: 'basic',
        title: '基本情報',
        fields: [FieldDefinition(field: 'code', label: 'コード', required: true)],
      ),
      WizardStepDefinition(
        id: 'address',
        title: '住所',
        fields: [FieldDefinition(field: 'zip', label: '郵便番号')],
      ),
    ],
  );

  final index = ScreenIndex.ofPages([orderSearch, customerMaster, wizard]);

  group('one line per screen', () {
    test('says what the screen is and how big it is', () {
      final brief = briefOf(orderSearch);
      expect(brief.what, '照会（読み取り専用）');
      expect(brief.line, contains('受注照会（order_search）'));
      expect(brief.line, contains('条件 2'));
      expect(brief.line, contains('列 2'));
      expect(brief.line, contains('ボタン 1'));
      expect(brief.line, contains('orderRepository から'));
    });

    test('counts steps for a wizard, not one folded form', () {
      final brief = briefOf(wizard);
      expect(brief.counts['steps'], 2);
      expect(brief.counts['fields'], 2);
      expect(brief.line, contains('ステップ 2（項目 2）'));
    });

    test('required fields are counted, but do not inflate the size', () {
      final brief = briefOf(customerMaster);
      expect(brief.counts['required'], 1);
      // 列 1 + 枠 1 + 項目 2 = 4（必須の 1 は項目に含まれているので数えない）。
      expect(ScreenEntry.of(customerMaster).size, 4);
    });
  });

  group('searching the index', () {
    test('every word must match (and case is ignored)', () {
      expect(
        index.search('顧客 検索').map((one) => one.id),
        // マスタ保守は「検索して一覧に出し…」なので検索でも当たる。
        containsAll(['customer_master']),
      );
      expect(index.search('ORDERREPOSITORY').map((one) => one.id), ['order_search']);
      expect(index.search('顧客 帳票'), isEmpty);
    });

    test('finds a screen by a label on it, and by a field name', () {
      expect(index.search('金額').map((one) => one.id), ['order_search']);
      expect(index.search('orderNo').map((one) => one.id), ['order_search']);
    });

    test('finds a screen by a button label', () {
      expect(index.search('CSV出力').map((one) => one.id), ['order_search']);
    });

    test('finds a wizard by a step title', () {
      expect(index.search('住所').map((one) => one.id), ['customer_wizard']);
    });

    test('no words means every screen', () {
      expect(index.search(null).length, 3);
      expect(index.search('   ').length, 3);
    });
  });

  group('the index itself', () {
    test('is sorted so the same input gives the same index', () {
      expect(index.screens.map((one) => one.id), [
        'customer_master',
        'customer_wizard',
        'order_search',
      ]);
    });

    test('biggest screens first, for finding where the work is', () {
      expect(index.bySize().first.id, 'order_search');
    });

    test('renders as a table with the columns lined up', () {
      final text = renderScreenIndex(index.screens);
      expect(text, startsWith('画面 3 枚:'));
      final lines = text.split('\n').skip(1).toList();
      // 画面名の位置が3行で揃っている（全角を2として詰めている）。
      final at = lines.map((line) => line.indexOf('顧客マスタ')).toList();
      expect(lines.length, 3);
      expect(at.first, greaterThan(0));
    });

    test('says so when nothing matches, instead of printing an empty table', () {
      expect(renderScreenIndex(index.search('存在しない画面')), contains('当てはまる画面はありません'));
    });

    test('an app indexes each of its screens', () {
      const app = AppDefinition(
        id: 'sales',
        title: '販売管理',
        pages: [orderSearch, customerMaster],
      );
      expect(ScreenIndex.ofApp(app, file: 'sales_app.yaml').screens.length, 2);
      expect(
        ScreenIndex.ofApp(app, file: 'sales_app.yaml').screens.first.file,
        'sales_app.yaml',
      );
    });

    test('is machine readable too (a screen picker can read it)', () {
      final json = ScreenEntry.of(orderSearch).toJson();
      expect(json['id'], 'order_search');
      expect(json['kind'], 'search');
      expect(json['what'], '照会（読み取り専用）');
      expect(json['repository'], 'orderRepository');
      expect((json['words']! as List).contains('金額'), isTrue);
    });
  });

  group('reading the parts of a page', () {
    test('a kind without that part answers, rather than throwing', () {
      expect(customerMaster.searchArea, isNull);
      expect(wizard.tableArea, isNull);
      expect(wizard.formArea, isNull, reason: 'ウィザードの項目は steps 側で数える');
      expect(wizard.steps.length, 2);
      expect(orderSearch.formArea, isNull);
      expect(orderSearch.pageActions.length, 1);
    });

    test('every kind reports its own type', () {
      expect(pageKindOf(orderSearch), PageKinds.search);
      expect(pageKindOf(customerMaster), PageKinds.master);
      expect(pageKindOf(wizard), PageKinds.wizard);
    });
  });
}
